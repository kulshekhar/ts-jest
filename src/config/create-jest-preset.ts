import { rootLogger } from '../util/logger'

const logger = rootLogger.child({ namespace: 'jest-preset' })

export interface TsJestPresets {
  transform: Record<string, string>
  testMatch?: string[]
  moduleFileExtensions?: string[]
}

export interface CreateJestPresetOptions {
  allowJs?: boolean
}

export function createJestPreset(
  { allowJs = false }: CreateJestPresetOptions = {},
  from: jest.InitialOptions = {},
): TsJestPresets {
  logger.debug({ allowJs }, 'creating jest presets', allowJs ? 'handling' : 'not handling', 'JavaScript files')
  return {
    transform: {
      ...from.transform,
      [allowJs ? '^.+\\.[tj]sx?$' : '^.+\\.tsx?$']: 'ts-jest',
    },
    ...(from.testMatch ? { testMatch: from.testMatch } : undefined),
    ...(from.moduleFileExtensions ? { moduleFileExtensions: from.moduleFileExtensions } : undefined),
  }
}
