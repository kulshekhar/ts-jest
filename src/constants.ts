export const LINE_FEED = '\n'
export const TS_TSX_REGEX = /\.tsx?$/
export const JS_JSX_REGEX = /\.jsx?$/
export const DECLARATION_TYPE_EXT = '.d.ts'
// `extensionsToTreatAsEsm` only accepts `.ts`, `.tsx` and `.jsx`. `.js`, `.cjs`, `.mjs` will throw error
export const TS_EXT_TO_TREAT_AS_ESM = ['.ts', '.tsx']
export const JS_EXT_TO_TREAT_AS_ESM = ['.jsx']
export const ALL_ESM_OPTIONS_ENABLED = {
  supportsDynamicImport: true,
  supportsExportNamespaceFrom: true,
  supportsStaticESM: true,
  supportsTopLevelAwait: true,
}
/**
 * @internal
 * See https://jestjs.io/docs/en/configuration#testmatch-arraystring
 */
export const DEFAULT_JEST_TEST_MATCH = ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)']
