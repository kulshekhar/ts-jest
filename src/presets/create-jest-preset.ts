import type { Config } from '@jest/types'

import { rootLogger } from '../utils/logger'

const logger = rootLogger.child({ namespace: 'jest-preset' })

export type TsJestPresets = Pick<
  Config.InitialOptions,
  'extensionsToTreatAsEsm' | 'moduleFileExtensions' | 'transform' | 'testMatch'
>

export function createJestPreset(allowJs = false, extraOptions: Config.InitialOptions = {}): TsJestPresets {
  logger.debug({ allowJs }, 'creating jest presets', allowJs ? 'handling' : 'not handling', 'JavaScript files')

  const { extensionsToTreatAsEsm, moduleFileExtensions, testMatch } = extraOptions

  return {
    ...(extensionsToTreatAsEsm ? { extensionsToTreatAsEsm } : undefined),
    ...(moduleFileExtensions ? { moduleFileExtensions } : undefined),
    transform: {
      ...extraOptions.transform,
      [allowJs ? (extensionsToTreatAsEsm?.length ? '^.+\\.m?[tj]sx?$' : '^.+\\.[tj]sx?$') : '^.+\\.tsx?$']: 'ts-jest',
    },
    ...(testMatch ? { testMatch } : undefined),
  }
}
