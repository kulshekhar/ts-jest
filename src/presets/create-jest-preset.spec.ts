import { createJestPreset } from './create-jest-preset'
import { JS_EXT_TO_TREAT_AS_ESM, TS_EXT_TO_TREAT_AS_ESM } from '../constants'

describe('create-jest-preset', () => {
  const baseExtraOptions = {
    testMatch: ['foo'],
    moduleFileExtensions: ['bar'],
    transform: { foo: 'bar' },
  }

  test.each([
    {
      allowJs: undefined,
      extraOptions: undefined,
    },
    {
      allowJs: false,
      extraOptions: undefined,
    },
    {
      allowJs: true,
      extraOptions: undefined,
    },
    {
      allowJs: true,
      extraOptions: {},
    },
    {
      allowJs: false,
      extraOptions: {},
    },
    {
      allowJs: false,
      extraOptions: baseExtraOptions,
    },
    {
      allowJs: true,
      extraOptions: {
        ...baseExtraOptions,
        extensionsToTreatAsEsm: [...JS_EXT_TO_TREAT_AS_ESM, ...TS_EXT_TO_TREAT_AS_ESM],
      },
    },
  ])('should return correct preset', (data) => {
    expect(createJestPreset(data.allowJs, data.extraOptions)).toMatchSnapshot()
  })
})
