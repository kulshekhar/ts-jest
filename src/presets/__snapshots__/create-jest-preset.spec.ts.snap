// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-jest-preset should return correct preset 1`] = `
Object {
  "transform": Object {
    "^.+\\\\.tsx?$": "ts-jest",
  },
}
`;

exports[`create-jest-preset should return correct preset 2`] = `
Object {
  "transform": Object {
    "^.+\\\\.tsx?$": "ts-jest",
  },
}
`;

exports[`create-jest-preset should return correct preset 3`] = `
Object {
  "transform": Object {
    "^.+\\\\.[tj]sx?$": "ts-jest",
  },
}
`;

exports[`create-jest-preset should return correct preset 4`] = `
Object {
  "transform": Object {
    "^.+\\\\.[tj]sx?$": "ts-jest",
  },
}
`;

exports[`create-jest-preset should return correct preset 5`] = `
Object {
  "transform": Object {
    "^.+\\\\.tsx?$": "ts-jest",
  },
}
`;

exports[`create-jest-preset should return correct preset 6`] = `
Object {
  "moduleFileExtensions": Array [
    "bar",
  ],
  "testMatch": Array [
    "foo",
  ],
  "transform": Object {
    "^.+\\\\.tsx?$": "ts-jest",
    "foo": "bar",
  },
}
`;

exports[`create-jest-preset should return correct preset 7`] = `
Object {
  "extensionsToTreatAsEsm": Array [
    ".jsx",
    ".ts",
    ".tsx",
  ],
  "moduleFileExtensions": Array [
    "bar",
  ],
  "testMatch": Array [
    "foo",
  ],
  "transform": Object {
    "^.+\\\\.m?[tj]sx?$": "ts-jest",
    "foo": "bar",
  },
}
`;
